### PR TITLE
Timers related function, set/clear Immediate/Timeout/Interval

### PR DESCRIPTION
**Pull request by zhanghuanrong**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -33,7 +33,7 @@
     "benchmark": "node benchmark/bench.js",
     "install": "node scripts/install.js",
     "prepare": "tsc -p lib && tsc -p test && tsc -p benchmark",
-    "test": "mocha test --recursive",
+    "test": "mocha test -g \"^((?!napajs/timers).)*$\" --recursive && mocha test -g \"^napajs/timers\"",
```
Comment (fs-eire) [inline package.json]: What does this part do

File: .travis.yml [hunk 2]
```diff
@@ -56,7 +56,7 @@ matrix:
     - COMPILER_OVERRIDE="CXX=g++-5 CC=gcc-5"
   # Node LTS (8.x)      OS X (El Capitan)   AppleClang 7.3
   - os: osx
-    node_js: '8'
+    node_js: 8
```
Comment (fs-eire) [inline .travis.yml]: should revert this file change

File: src/module/core-modules/napa/timer-wrap.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+void TimerWrap::Reset() {
+    _timer.reset();
+}
+
+napa::zone::Timer& TimerWrap::Get() {
+    return *_timer;
+}
+
+static void dummyWeakCallback(const v8::WeakCallbackInfo<int>& data) {
```
Comment (fs-eire) [inline src/module/core-modules/napa/timer-wrap.cpp]: UpperCamelCase for function names

File: src/module/core-modules/napa/timer-wrap.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+std::shared_ptr<napa::zone::CallbackTask> buildTimeoutTask(
```
Comment (fs-eire) [inline src/module/core-modules/napa/timer-wrap.cpp]: Let's define class CallbackTask inside this file and keep it local. There is no reason to introduce a public null task implementation.

File: src/module/core-modules/napa/timer-wrap.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                    parameters.emplace_back(v);
+                }
+                cb->Call(context, context->Global(), static_cast<int>(parameters.size()), parameters.data());
+
+                Local<Number> interval = Local<Number>::Cast(timeout->Get(String::NewFromUtf8(isolate, "_repeat")));
+                bool isInterval = (interval->Value() >= 1);
+                if (isInterval) {
+                    auto jsTimer = NAPA_OBJECTWRAP::Unwrap<TimerWrap>(
+                        Local<Object>::Cast(timeout->Get(String::NewFromUtf8(isolate, "_timer"))));
+                    jsTimer->Get().Start(); //re-arm
```
Comment (fs-eire) [inline src/module/core-modules/napa/timer-wrap.cpp]: revise the comments

File: src/module/core-modules/napa/timer-wrap.cpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+                    needDestroy = false;
+                }
+            }
+
+            if (needDestroy) {
```
Comment (fs-eire) [inline src/module/core-modules/napa/timer-wrap.cpp]: Add comments for object and context life-cycle management

File: src/module/core-modules/napa/timer-wrap.cpp [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    }
+    auto scheduler = zone->GetScheduler().get();
+    if (scheduler == nullptr) {
+        throw new std::runtime_error("Null scheduler encountered!");
+    }
+    auto workerId = static_cast<WorkerId>(
+        reinterpret_cast<uintptr_t>(WorkerContext::Get(WorkerContextItem::WORKER_ID)));
+
+    Local<Object> timeout = Local<Object>::Cast(args[0]);
+    auto sharedTimeout = std::make_shared<Persistent<Object>>(isolate, timeout);
```
Comment (fs-eire) [inline src/module/core-modules/napa/timer-wrap.cpp]: Suggested to use persistentTimeout / persistentContext

File: src/zone/schedule-phase.h [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+#include <stdint.h>
+
+namespace napa {
+namespace zone {
+
+    // Define the phase (like priority related type) of tasks. 
+    // To be used mainly by schduler and worker to schedule its tasks.
+    enum class SchedulePhase : uint32_t {
+        kDefaultPhase = 0,
```
Comment (fs-eire) [inline src/zone/schedule-phase.h]: remove the prefix k for enumerable definition

File: src/module/core-modules/napa/timer-wrap.cpp [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+    auto context = isolate->GetCurrentContext();
+    auto sharedContext = std::make_shared<Persistent<Context>>(isolate, context);
+
+    Local<Number> after = Local<Number>::Cast(timeout->Get(String::NewFromUtf8(isolate, "_after")));
+    std::chrono::milliseconds msAfter{static_cast<int>(after->Value())};
+    auto sharedTimer = std::make_shared<napa::zone::Timer>(
+        [sharedTimeout, sharedContext, scheduler, workerId]() {
+            auto timerCallbackTask = buildTimeoutTask(sharedTimeout, sharedContext);
```
Comment (fs-eire) [inline src/module/core-modules/napa/timer-wrap.cpp]: Should we create task object here? Seems it will be executed in the timer's scheduler thread.

File: src/zone/timer.cpp [hunk 10]
```diff
@@ -74,30 +74,30 @@ bool TimersScheduler::StartMainLoop() {
...
+
+                if (timers[expiredTimer.index].active) {
+                    timers[expiredTimer.index].active = false;
+
                     try {
-                        // Fire the callback.
-                        // The callback is assumed to be very fast as it is meant to dispatch to appropriate
-                        // callback queues.
-                        timers[topTimer.index].callback();
+                        // Fire the callback. The callback is assumed to be
```
Comment (fs-eire) [inline src/zone/timer.cpp]: Revise the comments here.

The comments are mainly about **why the change should be done in this commit**, but not explain the logic and behavior.

Also, this will not be fast since the new code tries to create a task object in `TimerWrap::SetTimersCallback`.

File: src/zone/timer.cpp [hunk 11]
```diff
@@ -74,30 +74,30 @@ bool TimersScheduler::StartMainLoop() {
...
                 }
-
-                // Timer expired, so it's no longer active.
-                activeTimers.pop();
             }
             else {
-                // Wait for timer expiration.
-                cv.wait_until(lock, topTimer.expirationTime, [&topTimer]() {
-                    return topTimer.expirationTime <= std::chrono::high_resolution_clock::now();
+                // Wait for timer expiration. Stop waiting if new urgent active timer is arm-ed.
```
Comment (fs-eire) [inline src/zone/timer.cpp]: is 'arm' a standard phrase? I got quite confused by it.

File: src/zone/worker.cpp [hunk 12]
```diff
@@ -132,15 +140,23 @@ void Worker::WorkerThreadFunc(const settings::ZoneSettings& settings) {
...
-            if (_impl->tasks.empty()) {
+            if (_impl->tasks.empty() && _impl->immediateTasks.empty()) {
                 _impl->idleNotificationCallback(_impl->id);
 
                 // Wait until new tasks come.
-                _impl->hasTaskEvent.wait(lock, [this]() { return !_impl->tasks.empty(); });
+                _impl->hasTaskEvent.wait(
+                    lock, 
+                    [this]() { return !(_impl->tasks.empty() && _impl->immediateTasks.empty()); });
             }
```
Comment (fs-eire) [inline src/zone/worker.cpp]: Add comments describing the new task draining behavior

File: test/timer-test.ts [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+    const NUMBER_OF_WORKERS = 3;
+    const kTaskGroupCount = 3;
+    let zone = napa.zone.create('zone', { workers: NUMBER_OF_WORKERS });
+
+    describe("napajs/timers", () => {
+    
+        describe("setImmediate/clearImmediate", function() {
+            let promises: Promise<napa.zone.Result>[] = [];
+            for (let groupId = 0; groupId < kTaskGroupCount; groupId++) {
+                let res = zone.execute('./timer-test', 'setImmediateTest', [groupId]);
```
Comment (fs-eire) [inline test/timer-test.ts]: create another file

File: src/zone/timer.cpp (hunk 11 above)
Comment (zhanghuanrong) [reply src/zone/timer.cpp]: It is quite common used especially event loop

File: src/module/core-modules/napa/timer-wrap.cpp (hunk 9 above)
Comment (zhanghuanrong) [reply src/module/core-modules/napa/timer-wrap.cpp]: ?? Task currently is the only existing way to be scheduled.

File: src/zone/schedule-phase.h (hunk 8 above)
Comment (zhanghuanrong) [reply src/zone/schedule-phase.h]: I saw this style in code. Where is the guide line?

File: .travis.yml (hunk 2 above)
Comment (zhanghuanrong) [reply .travis.yml]: yes, need revert it

File: package.json (hunk 1 above)
Comment (zhanghuanrong) [reply package.json]: other test affect result of the schedule precision. So split timer related.

File: test/timer-test.ts (hunk 13 above)
Comment (zhanghuanrong) [reply test/timer-test.ts]: I thinks this is same.

Comment (zhanghuanrong) [reply test/timer-test.ts]: And it could let people notice some tricky part of current implementation.

File: src/module/core-modules/napa/timer-wrap.cpp (hunk 5 above)
Comment (zhanghuanrong) [reply src/module/core-modules/napa/timer-wrap.cpp]: add some.

File: src/module/core-modules/napa/timer-wrap.cpp (hunk 7 above)
Comment (zhanghuanrong) [reply src/module/core-modules/napa/timer-wrap.cpp]: I'd like to keep it. As shared in JS and V8 also mean persistent to some extent, and also the shared_ptr semantic here.

File: src/zone/schedule-phase.h (hunk 8 above)
Comment (zhanghuanrong) [reply src/zone/schedule-phase.h]: fixed.

File: src/zone/worker.cpp (hunk 12 above)
Comment (zhanghuanrong) [reply src/zone/worker.cpp]: done!

File: src/zone/timer.cpp (hunk 10 above)
Comment (zhanghuanrong) [reply src/zone/timer.cpp]: revert back comments. Also create task is still fast enough. it just create several object.

File: src/module/core-modules/napa/timer-wrap.cpp (hunk 4 above)
Comment (zhanghuanrong) [reply src/module/core-modules/napa/timer-wrap.cpp]: Hide CallbackTask inside the only related cpp file.

File: src/module/core-modules/napa/timer-wrap.cpp (hunk 9 above)
Comment (zhanghuanrong) [reply src/module/core-modules/napa/timer-wrap.cpp]: To be clear, it just create task, but not run it, when timer call its call back.

File: src/module/core-modules/napa/timer-wrap.cpp [hunk 14]
```diff
@@ -0,0 +1,227 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+// See: https://groups.google.com/forum/#!topic/nodejs/onA0S01INtw
+#ifdef BUILDING_NODE_EXTENSION
```
Comment (helloshuangzi) [inline src/module/core-modules/napa/timer-wrap.cpp]: do you really need this here?

Comment (zhanghuanrong) [reply src/module/core-modules/napa/timer-wrap.cpp]: Yeah. it is no use here.
